torch/_dynamo/variables/nn_module.py

```diff
@@ -303,33 +303,29 @@ def call_method(
         key = self.module_key
         module = tx.output.get_submodule(key)
 
-        if name == "__call__":
-            # TODO(whc)  do we really need this special case?
+        if name == "_call_impl":
+            # Example: `self.layer.__call__(x)`
+            # This is used for explicit calling `__call__` in a forward function.
+            # Dynamo inlines `__call__`, includes hooks.
             return self.call_function(tx, args, kwargs)
-        elif name == "forward":
-            # TODO(whc)
-            # This is the old special case moved to a new place.  (copy from call_function below)
-            # Old behavior: we'd route "forward" meth call to 'call_function', which inlined forward.
-            # New behavior: since call_function now hits '__call__', forward would fall through to 'wrap_proxy' below,
-            # instead of being inlined.  What should we do about this?
-            #   1) all methods get inlined now at the bottom of this call_method, instead of put into the graph as calls
-            #   2) we maintain this special case just for forward
-            assert self.source, (
-                "Must provide a valid source in order to inline, "
-                "since inlined function may have default args which must be guarded."
-            )
-            fn = module.forward.__func__
-            assert istype(fn, types.FunctionType)
-            options["source"] = AttrSource(
-                AttrSource(self.source, "forward"), "__func__"
-            )
-            args = [self] + args
+        elif name == "forward" and is_allowed(module.__class__):
+            # Example: `self.layer.forward(x)`
+            # This is used for explicit calling `forward` in a forward function,
+            # and self.layer is allowed module.
+            # Dynamo puts `call_module` node in FX, doesn't trigger hooks.
+            # For not allowed module, it has been inlined before getting here.
+            from .builder import wrap_fx_proxy
 
-            return tx.inline_user_function_return(
-                variables.UserFunctionVariable(fn, **options),
-                args,
-                kwargs,
-            )
+            with self.record_nn_module_stack(tx, module):
+                return wrap_fx_proxy(
+                    tx=tx,
+                    proxy=tx.output.create_proxy(
+                        "call_module",
```

Comment:
I thought we agreed that '.forward' should not route to 'call_module'?

Reply:
Oh, I think I have misunderstanding for this, let me clarify more:
* If users write ```self.layer(x)``` and self.layer is allowed module:
  * w/ hooks: inlining hooks calls and put ```call_module``` into FX.
  * w/o hooks: put ```call_module``` into FX.
* If users write ```self.layer.forward(x)``` and self.layer is allowed module:
  * w/ hooks: N/A
  * w/o hooks: ? <--- This is the change in this PR, I'd say put ```call_module``` into FX is more consistent, and I think we should try to use ```call_module``` if we can.

Reply:
ah.. i see. I don't think we should map `self.layer.forward(x)` to  `call_module` ever.  It is also a user pattern to register a hook later (after compile).  If the user turns on 'enable hooks guards', then dynamo should observe the new hook and recompile.

Reply:
Addressed offline, thanks for educating me how AOT autograd handle ```call_module```.